### ISSUE TITLE
Query Operator Functionality

### ISSUE THREAD

**Issue by kazetsukaimiko** [ojplg/hrorm#3]
Quick mock-up of a fluent query interface. Not sure how you'd do things like date ranges otherwise.
https://github.com/kazetsukaimiko/hrorm/blob/fluent-query/src/test/java/org/hrorm/FluentTest.java

Each Conditional<ENTITY> could have a StringBuilder and assemble their own query in a toString() or toSQL() method, conceivably. Or, you could have something like SqlBuilder create a query by reading into a Compound<ENTITY.

There are good reasons to want to avoid something like this, just pitching the idea.

**Comment by ojplg** [comment]
Yeah, I have thought about this issue. I am not at all sure what a good solution is.

As you know, hrorm only supports equality matching in querying. And does nothing involving logical combinations of AND-ing and OR-ing.

To me those problems are orthogonal. So, there is an issue of matching, where matching can mean less than or greater than (for numerics and timestamps) and LIKE queries for Strings. Then there are issues of AND and OR and other complexities.

The second issue seems less worth solving to me. You can do UNIONs or ANDs in code. Just glob the lists together. If you really need a specialized select with a hairy where clause, the hrorm ethos is to say: that's what JDBC is for. Use hrorm for the basics, and if you have a few other things, well, just write the SQL and parse the results, it's not that hard.

Perhaps it would be useful to expose the result set parsing and statement management that hrorm includes, but I am not sure how to do that.

The case of the less than and like clauses is more ambiguous to me. I just don't have a great solution.

I would like to have things follow the templating approach though. A much simpler solution to what you propose is to overload the `selectManyByColumns` method on `Dao` to look like this:

`selectManyByColumns(T template, Map<String, Operation> columns)`

Where `Operation` is an enumeration with values include LessThan, LessThanOrEquals, Like, GreaterThan, GreaterThanOrEquals.

That would solve one part of the problem, but not the other. But like I said, I'm not sure it's hrorm-y in spirit to be an all-purpose SQL builder.

**Comment by kazetsukaimiko** [comment]
I guessed as much, and thinking about and/or behavior specifics had me second-guessing my proposed approach.

There's certainly much less of a need for OR than AND.

A slight variant, 
`selectManyByOperations(List<Operation> ops) `
can even work with ranges, if Operation is a typed object and contains templates/values. If Operation is an enum, you couldn't do more than one comparison per column.

```
new LessThan(String column, T template)
new LessThanOrEquals(String column, T template)
new equals(String column, T template)
new notEquals(String column, T template)
new Range(String column, T lowerTemplate, T upperTemplate)
new In(String column, List<T> templates)
```

**Comment by ojplg** [comment]
Really, the way hrorm works is that if you specify multiple values in a template, there's an AND added to the where clause. You could overload the method to have an OR variant, but I don't think that buys much. I think the AND is more useful, and, as I said, the OR can be done by running multiple queries and globbing. (Though you can get the same record twice that way.) And if an OR variant is added, you still cannot get arbitrary nestings of booleans that SQL allows.

Hrorm does not support any kind of intersection or having clauses. SQL is big. Hrorm is small.

Your Operation as a class idea is possible, but becomes trickier to implement. An enum just translates one-to-one to a SQL operator. Also, it perhaps is less-like the templating approach. Though, it's a bit unclear. Are you imagining the T in your examples as being the Entity type, or the Field type?

Another thing that bugs me is that when you do the select you have to use the COLUMN names. That's not really the right thing: you should use the field name in the class, right? But, the column name makes generating the SQL easier. And, using the field name starts down the road toward reflection. You could use the accessor function name, but ... that seems weird. I've never really been happy with this.

**Comment by kazetsukaimiko** [comment]
If you changed how fields are declared in DaoBuilder, you may end up with something more to your tastes, at least on the query-by-field side of things by adding your column definitions as static-final properties on your entity classes (or anywhere, really).

Take your Child class under tests for example:
`public static final Column<Child, Long> ID = DataColumn.primaryKey("id", Child::getId, Child::setId);
// For each field you care about`

Then when building:
```
DaoBuilder<Child> childDaoBuilder = new DaoBuilder<>("child", Child::new)
            .field(Child.ID)
            // ...
```

Now, whenever you want to reference that column, you have all the type information:
`Child childById = childDao.findOneBy(Child.ID, 1L)`

And to answer your question, I was thinking T = a template entity as you have used, modelling after the pattern you've created.

It would also be fairly easy to implement very basic OR using the Operation class model I described:
new OROperation(Operation... operation)

But it suffers from similar criticisms to those you had prior.

**Comment by kazetsukaimiko** [comment]
So, I started exploring this and found something strange. So far, I can see no justification for the *BUILDER generics on DaoDescriptor and others. It seems like BUILDER and ENTITY are interchangeable, and if ENTITYBUILDER really was as it is described in DaoDescriptor:
`
 * @param <ENTITYBUILDER> The type of object that can build an <code>ENTITY</code> instance.`

Then this doesn't make sense to me:
```
    /**
     * The mechanism to use to instantiate a new instance of type <code>ENTITY</code>,
     * generally a no-argument constructor of the class.
     *
     * @return A function pointer to the instantiation mechanism
     */
    Supplier<ENTITYBUILDER> supplier();

```
At the very least if ENTITYBUILDER is some kind of Supplier/entity construction class (then why need the extra type, why not just use Supplier<ENTITY>?), it should have some bounds to make it usable.

I have another branch where I tried exploring the static metamodel approach I described above with Child.ID, as a part of that I needed to take a closer look at the generics. I started with a new type Field<ENTITY, TYPE>, as allowing the construction of Column<...> outside of its current context seemed difficult and unweildy.

```
public interface Field<ENTITY, TYPE> {
    /**
     * Name of the column backing this field.
     * @return Name of the column in the database table.
     */
    String getName();

    /**
     * Getter for the field on the entity.
     * @return Function representing the getter on the entity.
     */
    Function<ENTITY, TYPE> getGetter();

    /**
     * Setter for the field on the entity.
     * @return BiConsumer representing the setter on the entity.
     */
    BiConsumer<ENTITY, TYPE> getSetter();
}
```

Then AbstractColum's constructor changes to:

`public AbstractColumn(String prefix, Field<ENTITY, TYPE> fieldDefinition, boolean nullable)`

I think the nice things about this approach (compile-time type safety, no reflection, its completely optional for existing functionality) make it a good candidate. Once I have something working, I'll let you know.

**Comment by ojplg** [comment]
Take a look at the `IndirectDaoBuilder`. Take a look at the way the `ImmutableThingTest` works. You will understand why there is a difference between an entity and a builder.

**Comment by kazetsukaimiko** [comment]
That "OHHHHH" moment....

I can't say there's a better solution for supporting immutable entities/final fields without reflection, I'm really going to have to think about it. I have a hunch you're way ahead of me. What is the goal there? 

The BUILDER generics really make it difficult to do what I'm suggesting cleanly. 

My suggestion to solving immutable in a less invasive way:
```

DaoBuilder<Mutable> mutableDaoBuilder = new DaoBuilder<Mutable>(/* ... */)
  .fieldDefinitions();
Dao<Mutable> mutableDao = mutableDaoBuilder.build(connection);

Function<Mutable, Immutable> mappingFunction = (mutable) -> new Immutable(mutable);

Dao<Immutable> immutableDao = mutableDao.map(mappingFunction);

```
Then you have a very thin wrapper: 
`DaoMapper<ENTITY, MAPPEDENTITY> implements Dao<MAPPEDENTITY> `
that applies the mapping function to the passed in `Dao<ENTITY>`.

"Mutable" could conceivably be ImmutableThing.ImmutableThingBuilder, and mappingFunction could be ImmutableThing.ImmutableThingBuilder::build.

**Comment by kazetsukaimiko** [comment]
MappingFunction would need to be a Converter<Mutable, Immutable> to be seamless.

**Comment by kazetsukaimiko** [comment]
By the way, Is there a quick way to get a custom query to execute in HRORM as it is on the keyless branch? I have method stubs on my service layer for the date range queries I really need right now, and it'd help to be able to get something working until we can vet a proper enhancement post keyless.

**Comment by ojplg** [comment]
There is no way to inject queries into hrorm or do ranges or less than yet.

**Comment by ojplg** [comment]
I just pushed some code to master that has a new mechanism for doing selects with comparisons.
Right now, it only supports a 'LIKE' operator for strings and <,>,<=,>= for numbers.

Extending the work to cover a range of numbers should be easy, but i have not done it yet.

Covering dates is a big problem though. The issue is SQL syntax.

Hrorm only uses ANSI SQL, and almost all date and time manipulation is database engine dependent. There's very little that works on Oracle, SQLServer, Postgres, MySQL, ...

I am not sure what I want to do about that.

Take a look at what I have so far though, and see if that looks promising.

**Comment by kazetsukaimiko** [comment]
Let me pull. I'll take a look.

As for timestamp syntax in the varying dialects, I'd suggest patterning your solution after `withConvertingStringColumn`, like a `withTemporalField(TEMPORALTYPE temporal)`. Let the implementer figure this out by giving them a direct way to say how to convert (whatever their timestamp field is- LocalDateTime, ZonedDateTime, etc) into a dialect-specific SQL expression you blindly place into the query at the right place, and then throw good exceptions when the implementer does something wrong.

And if you're really nice, you could even include a few in HRORM for them:
```
OracleDialectTimeConverter
MySQLDialectTimeConverter
MSSQLDialectTimeConverter // This uses DateTimeFormatter.ISO_DATETIME anyway
PostgresDialectTimeConverter
//... 
```

etc

**Comment by kazetsukaimiko** [comment]
I just went through and converted the internals to Stream entities, see the PR I just issued. Tomorrow I'm going to try and write out all the tests against your Operation functionality, starting in KeylessTest. I'll see if I can cover each Operation type, it should give me an idea of the limitations there.

**Comment by kazetsukaimiko** [comment]
First problem I'm seeing with Operator.LIKE:

```
Map<String, Operator> columnOperatorMap = new HashMap<>();
columnOperatorMap.put("string_column", Operator.LIKE);

// Without this, nothing is returned.
template.setStringColumn('%'+template.getStringColumn()+'%');

List<Keyless> fromDatabase = dao.selectManyByColumns(template, columnOperatorMap);
```

This may be your intention, to support LIKE for %value%, %value and value%.

**Comment by ojplg** [comment]
About the necessity of putting in ampersands when doing LIKE queries. I think that is the best way to do it. It's annoying, I admit, but what if I want to specify that the select is for things that start with FOO, not just things that contain FOO. If hrorm automatically sticks ampersands on the passed String, I am preventing users from making that distinction.

**Comment by ojplg** [comment]
I added code and tests for ranged select by columns.

There is some clean up remaining to be done, but I think it's basically working.

If this meets your needs, I think it's getting close to time to cut a publish a new release of hrorm.

**Comment by kazetsukaimiko** [comment]
I think this meets my immediate needs just fine. Ideally, there'd be OR support, but that's a separate effort if you want to pursue that. I can pretty easily split each side of the OR into two queries that do not overlap.

**Comment by kazetsukaimiko** [comment]
Renamed so that this issue can be closed at your convenience. If there's anything else in here you'd like to continue working on, I'll open another issue for it. Personally I kind of like the idea of supporting type-safe field queries through a static metamodel.

**Comment by ojplg** [comment]
It should be relatively easy to add another method (or methods) that links together the query criteria with ORs instead of ANDs. Of course that still leaves out the issue of mixing ANDs and ORs arbitrarily.

Your idea of a metamodel is valid. I think it's some work ... but a lot of ORMs go down that road and end up with a lot of code and still don't cover all you can do with SQL. It's a hard decision. What we have now in hrorm is certainly growing in complexity.

It's hard to know where to draw the line. One advantage tools like JDBI have is that you know there is portability, since you can just cut-and-paste the SQL you wrote yourself into a new library or database GUI or whatever. These new hrorm things might be expressive or might now, but they are certainly completely useless to any other library or database interface. I think metamodels suffer from that problem even more.

Anyway, I will look at adding the ORs some time soon.

Once this is settled down, we should cut a new release of hrorm, as I said before.

**Comment by kazetsukaimiko** [comment]
Agreed on the release- the sooner the better!

For the metamodel, I can mock something up. I was thinking that its mostly a cosmetic change- you either pass in a Field at DAO build (which the Dao/Builder now stores as a field instead of separate column name / setter/ getter), or you call existing methods that internally build those Field objects anyway (if you don't even care enough to use the feature). Then we could add convenience methods on the DAO that chain using the type-safety to build a `Map<String, Operator>` and template for invocation.
```

KeylessDao<Keyless> dao;
List<Keyless> dao
   .field(Keyless.INTEGER_COLUMN, Operator.EQUALS, 0L)
   .field(Keyless.STRING_COLUMN, Operator.LIKE, "somePrefix%")
   .findAll(); //Invokes existing DAO methods on the calling DAO by populating a template and Map<String, Operator> values
```

This makes having to set 'somePrefix%' much less ugly, no?

And there's no mandate to store any of these metadata fields on Keyless, or anywhere at all- its just there if you want Field query support.

**Comment by kazetsukaimiko** [comment]
Since Operator is a class and not an enum, we could have StringOperator ( LIKE ),  NumericOperator( LT/GT/LTE/GTE ) subclasses instead of making them the base Operator ( EQUALS, NOT_EQUALS ) instances. Then, Field can have three declarations to enforce which Operators are allowed for which types:
```

<E extends Number> T extends Field<ENTITY, E>> field(T numericField, NumericOperator operator, E fieldValue);
<E extends Temporal> T extends Field<ENTITY, E>> field(T temporalField, NumericOperator operator, E fieldValue);
<T extends Field<ENTITY, String>> field(T stringField, StringOperator operator, String fieldValue);

```
This will prevent:
`.field(Keyless.INTEGER_COLUMN, Operator.LIKE, 0L); // Operator.LIKE is still declared on Operator, but is a StringOperator type.`

And its easier to use while preventing misbehavior with badly build Map/columnname/operator combinations.

**Comment by kazetsukaimiko** [comment]
Uh oh, I just realized this isn't going to work. I need to do two comparisons against the same column, so the template model isn't going to work.

I need support for datefield < something AND datefield > somethingElse. 

For the time being I'm going to implement using:
datefield > earliest AND someotherqualifier = somevalue

That's a good enough approximation that I can get over the finish line without too huge of overhead until something better comes along.

**Comment by ojplg** [comment]
Take a look at the Operator.openRangeTo() and Operator.closedRangeTo() functions.

If I understand what you are saying, those should do what you want.

**Comment by kazetsukaimiko** [comment]
Since its a `Map<String, Operator>`, there's no way to specify more than one operation per column.  
SELECT * FROM LOG WHERE TIMESTAMP < ? AND TIMESTAMP > ?
? and ? being today, yesterday respectively.

You'd need something like a `List<Predicate>` to overcome this.

**Comment by ojplg** [comment]
Look at the code and the tests. It does exactly that.

**Comment by kazetsukaimiko** [comment]
Nevermind, I am an idiot! Those are exactly what I need. I think I recall that class being smaller/more minimalistic.

**Comment by ojplg** [comment]
It's possible that the class should be renamed, since Operator is only part of it.

Something like "PartOfAWhereClauseRelatedToAParticularColumn" but that does not suck.

This stuff is all new. There are better ideas out there, but we have to find them.

**Comment by kazetsukaimiko** [comment]
Assuming usage is something like:
```
Log template = new Log();
template.setsomeField(someValue);
template.setRequestDatetime(timeFrame.start);
Map<String, Operator> comparisons = new HashMap<>();
comparisons.put("somefield", Operator.EQUALS);
comparisons.put("request_datetime", Operator.openRangeTo(timeFrame.finish));
```

Yeah, I think this is clearly scope creep in **something like an** Operator class. No big deal, its the quick fix. Much appreciated.

**Comment by ojplg** [comment]
It kinda works, but it is limited to one case.

Your idea of building up queries with a meta language is more intuitive and more flexible. It's somewhat surprising how he Operator range methods work.

The problem with the meta language is that it starts to get to a lot of code. You end up supporting more and more things and it just gets bigger and bigger and more complicated and you never stop. And you still cannot support all of SQL.

So, then you got the JDBI route and just have people write the SQL themselves and cut and paste it into literal strings in Java. That obviously sucks in other ways.

With hrorm, I'm trying to strike a balance. Support the 80% of use cases with 20% of the code. (I hope.) But it's very much a matter of opinion what you should support and exactly how.

It's been very gratifying to work with you on this. This is functionality that hrorm really needs. I am not sure if this is a good solution yet.

Thanks for your contributions, wherever this goes from here.

**Comment by kazetsukaimiko** [comment]
Yeah, likewise on the working together. I've found myself often having to take a step back and remind myself how I'd react if someone was proposing changes to my code. Its your baby... I do hope we can continue to refine this together.

I guess I'll leave my opinion on "more code". Needless to say, as people take interest in your baby, its going to grow. Just the nature of the beast. I think winning contributions to any project almost invariably contain all of the following:
- Increase architectural consistency.
- Manage to simplify or de-duplicate internals somewhat.
- Adhere to your design ethos.
- Looks fun to hack.
- Increases capabilities throughout the stack. Its not a bolted-on limb.
- Don't change interface contracts, unless you're thinking major version bump.

Rather than emphasizing total library size, in most situations I'd say it is more beneficial to ensure that contributions bring lots of value, stick to your values and don't balloon cognitive complexity. More moving parts and lines of code is usually fine if it is still easy to understand at the end of the day.

I will agree, balance is hard to get right, its subjective and largely elusive. There's always something that falls short in some area and usually you're forced to compromise. At the end of the day, this is your repo. If you don't like something, you don't have to live with it- it can stay in the fork if the maintainer cares enough to maintain it. 

My overall impression of hrorm:
![hrorm](https://user-images.githubusercontent.com/3697327/52235730-f60e9d80-2892-11e9-99a3-29f119b008fc.jpg)

Before you get too upset about some of that, the same metrics for something like Hibernate, JPA (my typical stack):
![jpa_hibernate](https://user-images.githubusercontent.com/3697327/52235791-1c343d80-2893-11e9-8cb6-412ebd4f073d.jpg)

Not all that fun or quick to implement, and having spelunked into Spring/Hibernate stacks time and time again trying to figure out why I'm having problems I can tell you that the interfaces are obscure and clunky, the end result is anything but clear and you have to align the stars to get anything working at all.

I think you have a lot of room to fill out the right side without any real sacrifice on the left.

**Comment by ojplg** [comment]
I'm not at all upset by your assessment of hrorm. It's exactly what it was designed for: easy to use, limited capabilities. I think our opinion is spot on.

I agree that increasing the capabilities is good. But hrorm will always prioritize easy to use, even if that means things cannot be supported.

I did cut a new release of everything we have done up to now. Hrorm is now 0.6.0 thanks to you and your ideas. I think the new functionality for supporting selects is very useful, even if I think we can improve the design. I'm not sure how just yet.

I also updated the TODOs in the README. We've talked about these other ideas:
* Support selects that are connected with "OR" statements, not just "AND"
* Support different types, e.g. String GUIDs, for primary keys
* Add methods that allow for updates and deletes based on criteria other than primary key

Not sure if any of those are a priority for you.

I am closing this ticket as completed, but look forward to making more improvements to hrorm with you. Though of course, I have other things to do too.

Glad someone is finding it useful.
